### PR TITLE
SCP-2351: CekValEnv keyed by DeBruijnLvl, CEK evaluating Term DeBruijnIx

### PR DESCRIPTION
**Pull request by bezirg**
This PR changes the CekValEnv from a UniqueMap 
to a pair of:

1) IntMap CekValue, where the key signifies the debruijn level
2) the current debruijn level as seen from the side of the closure environment.

This allows to retain the fast IntMap as the underlying CEK datastore, while
switching the term under evaluation to use debruijn indices.

The most interesting change of this PR is in the dischargeCekValue/dischargeCekValEnv where
caution has to be taken to do the correct math for discharging the free variables.

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by bezirg** [comment]
Do not merge, plutus-becnhmark/validation/data/**/*.flat have to be updated because the DeBruijn index has changed from Natural to Word.

File: cabal.project [hunk 1]
```diff
@@ -187,11 +187,15 @@ source-repository-package
   location: https://github.com/input-output-hk/hedgehog-extras
   tag: 8bcd3c9dc22cc44f9fcfe161f4638a384fc7a187
 
+source-repository-package
```
Comment (michaelpj) [inline cabal.project:190]: I think drop this from the benchmarks. We don't need it for `Intmap+Levels`, and it gives us another awkward dependency.

File: plutus-tx-plugin/test/Lift/list.plc.golden [hunk 2]
```diff
@@ -2,46 +2,11 @@
...
-        )
-        (delay
-          (delay
-            (lam
-              case_GHC_Types_Nil_i0
-              (lam case_GHC_Types_Cons_i0 case_GHC_Types_Nil_i2)
-            )
-          )
-        )
+        (lam i0 (lam i0 (lam i0 [ [ (force i2) (con integer 1) ] (force i3) ])))
```
Comment (michaelpj) [inline plutus-tx-plugin/test/Lift/list.plc.golden:5]: This test output is unusuable now. Can we keep using named indices for this sort of stuff? We're not evaluating it.

**Review by michaelpj — COMMENTED** [review]
Couple of things that jump out, will review more later.

**Comment by johnalotoski** [comment]
This pr has been failing eval on hydra since 2021-06-22.  Disabling automatic eval and build with the `hydra-dont-build` label to avoid unnecessary hydra resource consumption.  Please remove this label from the PR when you want hydra to resume eval and builds.

**Pull request closed without merging**